refactor: further separate CLI logic from the API related functionality (see #117)

Rolv-Apneseth opened the pull request:
Continuation of #117.

As discussed, some refactors. You will probably want some more done here and that's OK, just let me know what to do.

One thing for sure is the `String` values in `api/check.rs` - which ones exactly would you like me to make `&str` or `Cow<'_, str>`?

Also, for `cli/mod.rs`, I could potentially remove the need for the manual call of `cmd.execute` for each subcommand using the [enum_dispatch](https://crates.io/crates/enum_dispatch) crate if you're OK with adding a dependency? Fine as it is anyway though in my opinion, as there aren't many subcommands.

jeertmans commented:
Hi @Rolv-Apneseth! I have quickly looked at your PR, it looks very good! I will give it a better look later, but let me answer your first questions.

> One thing for sure is the `String` values in `api/check.rs` - which ones exactly would you like me to make `&str` or `Cow<'_, str>`?

Well, I think `&str` should be enough, because we will never need to mutate (i.e., overwrite) the text strings, so keeping a reference should be fine. However, If I remember well, Clap makes this hard as you don't get reference to string read from the terminal, but owned string. So parsing text (either pure text or data annotation) from the terminal will required to be able to pass owned data, I fear. Using borrowed data is only useful for when reading files, as we may want to avoid allocating the full content of the files, and read by chunks instead.

So in conclusion, `Cow<'_, str`> may be the only solution (apart from `String`).

> 
> Also, for `cli/mod.rs`, I could potentially remove the need for the manual call of `cmd.execute` for each subcommand using the [enum_dispatch](https://crates.io/crates/enum_dispatch) crate if you're OK with adding a dependency? Fine as it is anyway though in my opinion, as there aren't many subcommands.

We can give it a try :-)
Dependencies should be fine, especially as we can keep it under the `cli` feature.

Rolv-Apneseth commented:
> I have quickly looked at your PR, it looks very good! I

Great!

> Well, I think `&str` should be enough, because we will never need to mutate (i.e., overwrite) the text strings, so keeping a reference should be fine. However, If I remember well, Clap makes this hard as you don't get reference to string read from the terminal, but owned string. So parsing text (either pure text or data annotation) from the terminal will required to be able to pass owned data, I fear. Using borrowed data is only useful for when reading files, as we may want to avoid allocating the full content of the files, and read by chunks instead.

Sorry, I meant which things in `check.rs` you actually wanted converted, but from this I'm assuming it's the `Request.text`? But I am confused how this would be achieved. Do we not need to allocate a `String` no matter what the source is? As for reading the file by chunks, do you want me to just make this use a `BufReader` and create and await requests one at a time instead of the way it is currently?

```rust
            let text = std::fs::read_to_string(filename)?;
            let requests = request
                .clone()
                .with_text(text.clone())
                .split(self.max_length, self.split_pattern.as_str());
            let response = server_client.check_multiple_and_join(requests).await?;
```

And for the `text.clone()` caused by `server_client.check_multiple_and_join`, I could make that function just return a `ResponseWithContext` instead of having it convert to `Response` automatically like it's doing, so we can pass the owned string in and get it back out of the response.

> We can give it a try :-)

I've added it in there - I think it fits nicely.

Rolv-Apneseth commented:
> I could make that function just return a `ResponseWithContext` instead of having it convert to `Response` automatically like it's doing, so we can pass the owned string in and get it back out of the response.

I've pushed some changes there for this, avoiding the need to clone the input text.

jeertmans commented:
> Sorry, I meant which things in `check.rs` you actually wanted converted, but from this I'm assuming it's the `Request.text`?

Yes, `Request.text` and `Request.data`'s inner fields. Mainly because one could create a vector of data annotations from a stream of Markdown tokens, which are usually `&str` of the initial source.

> But I am confused how this would be achieved.

I think using `Cow<'_, str>` should make the trick.

> Do we not need to allocate a `String` no matter what the source is?

Yes and no. Yes because `reqwest` will have to allocate a string for the HTML request, but no because we don't actually need the user to provide us an owned `String`. We don't care because `reqwest` will allocate anyway. So it's good to provide flexibility.

> As for reading the file by chunks, do you want me to just make this use a `BufReader` and create and await requests one at a time instead of the way it is currently?
> 
> ```rust
>             let text = std::fs::read_to_string(filename)?;
>             let requests = request
>                 .clone()
>                 .with_text(text.clone())
>                 .split(self.max_length, self.split_pattern.as_str());
>             let response = server_client.check_multiple_and_join(requests).await?;
> ```

Yes, this would be great. The split-text logic isn't really trivial to solve, because we want to avoid splitting text where it would break the meaning of a sentence, but also we cannot have long-running text either. I think working on a better split-logic can be a next step. For the moment, we can just use `BufReader` and read the whole file, and feed the server with multiple requests (if text is too large). We could also read the file by chunks, but I think this is a fair assumption to assume that any file we want to check should fit completely in the memory. 

> And for the `text.clone()` caused by `server_client.check_multiple_and_join`, I could make that function just return a `ResponseWithContext` instead of having it convert to `Response` automatically like it's doing, so we can pass the owned string in and get it back out of the response.

Yes, the end-goal would be to remove the complex logic from `Server` and perform multiple requests merging outside, so the server only has one method per HTTP request, nothing more.

Rolv-Apneseth commented:
Hey again. Had a lot of trouble with lifetimes and I could only get `'static` to work - not sure I love the whole solution, but if you havea look at the latest commit, is that what you had in mind?

jeertmans commented:
Hello @Rolv-Apneseth, rapidly looking at your commit, I think you just forgot to add the lifetime parameter in `Request -> Request<'source>` (I would prefer using `'source` over anonymous lifetime names like `'a`). This should explain why you could only use `'static`.

However, it makes little sense to have reference in responses, because `reqwest::RequestBuilder::send` returns an owned `Response`. We can keep owned `String` in that case :-)

This will replicate a similar behavior to that of `reqwest`. E.g., the builder takes a reference to be constructed `pub fn json<T: Serialize + ?Sized>(self, json: &T) -> RequestBuilder`, but the response returns an owned struct: `pub async fn json<T: DeserializeOwned>(self) -> Result<T>`.

jeertmans commented on src/api/check.rs:
```diff
@@ -607,15 +607,20 @@ impl Request {
-            .as_ref()
-            .ok_or(Error::InvalidRequest("missing text field".to_string()))?;
+    pub fn try_split(mut self, n: usize, pat: &str) -> Result<Vec<Self>> {
+        let text = mem::take(&mut self.text)
```
Here, one goal would be that `try_split` returns a reference to `&self`, so the signature should be as follows:

`pub fn try_split<'source>(&'source self, n: usize, pat: &str) -> Result<Vec<Self<'source>>>`

Note that we should not need to mutate the actual request, because we only take references.

Rolv-Apneseth commented:
I can have a go again but something was complaining about needing to live longer than `'static`. As for `Response`, that makes sense.

jeertmans commented:
> I can have a go again but something was complaining about needing to live longer than `'static`. As for `Response`, that makes sense.

No issue. If you can't make it compile, just push with `'source`, and I will try to fix it myself :-)

Rolv-Apneseth commented:
So this code:

```rust
#[cfg_attr(feature = "cli", derive(Args))]
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Hash)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct Request<'source> {
    /// The text to be checked. This or 'data' is required.
    #[cfg_attr(
        feature = "cli",
        clap(short = 't', long, conflicts_with = "data", allow_hyphen_values(true))
    )]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<Cow<'source, str>>,
```

Leads to this error:

```
error: lifetime may not live long enough
   --> src/api/check.rs:398:15
    |
391 | pub struct Request<'source> {
    |                    ------- lifetime `'source` defined here
...
398 |     pub text: Option<Cow<'source, str>>,
    |               ^^^^^^ requires that `'source` must outlive `'static`

error: lifetime may not live long enough
   --> src/api/check.rs:392:5
    |
391 | pub struct Request<'source> {
    |                    ------- lifetime `'source` defined here
392 |     /// The text to be checked. This or 'data' is required.
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ argument requires that `'source` must outlive `'static`
    |
    = note: this error originates in the macro `clap::value_parser` (in Nightly builds, run with -Z macro-backtrace for more info)
```

So I don't think it will work as it is. Should I take out `clap` from here and have a separate struct for the request arguments?

jeertmans commented:
> ^ requires that `'source` must outlive `'static`

Oh yeah, I remember why I struggled with `&str` back then ^^'

Creating a separate struct would work, but that means duplicating a lot of the code. I wonder if there is still a way to allow `'source` to take `'static` in the case of `App::parse`, but not in general... This isn't as easy as I imagined

Rolv-Apneseth commented:
> I wonder if there is still a way to allow `'source` to take `'static` in the case of `App::parse`, but not in general

Not that I can think of anyway. I'd say the options are to either stick with `String`, use `'static`, or a different struct. For the different struct, could an `impl From` keep fields in sync at least?

jeertmans commented:
> > I wonder if there is still a way to allow `'source` to take `'static` in the case of `App::parse`, but not in general
> 
> Not that I can think of anyway. I'd say the options are to either stick with `String`, use `'static`, or a different struct. For the different struct, could an `impl From` keep fields in sync at least?

Ok so some news: at the moment, it seems impossible to do that in one struct, see https://github.com/clap-rs/clap/issues/5773 and the related discussion, so I suggest we duplicated the structure: to original `Request`, with `Cow<'source, str>`, and the clap-compatible `RequestCli`, with `String`. Maybe we can simply avoid duplicating code with a macro rule. If you don't see how to do that, I think I can put have some time during the weekend for this.

Rolv-Apneseth commented:
I think it would have to be a procedural macro right? Can't see how to do that with a declarative one.

I have a bit of time today so I'll have a look but feel free to do it if you have a good idea of how it should be done

jeertmans commented:
I actually started some work on your branch today, I will hopefully have something working this afternoon :)

Rolv-Apneseth commented:
Oh nice, good luck

jeertmans commented:
Ok so it compiles with `--no-default-features` (without CLI), but I fear it isn't possible to conditionally compile the `#[deriver(Parser)]` based on lifetime `$lt == 'static`.

I am looking a possible crates that might provide useful tools to avoid rewriting most of the code.

jeertmans commented:
Unfortunately, I haven't been able to put more work on this PR this weekend, sorry for that.

I fear more and more that the copy/paste option would be the only way of solving this.

Rolv-Apneseth commented:
Hey - been busy moving this weekend. If I have any time in the coming days I can have a look for any crates that could solve this problem - if not, yes, copy paste might have to do for now

jeertmans commented:
> Hey - been busy moving this weekend. If I have any time in the coming days I can have a look for any crates that could solve this problem - if not, yes, copy paste might have to do for now

No issue! Yeah, maybe let's copy and paste (keeping the `'source` lifetime for the `api/click.rs` and use `String` in `cli/checks.rs`). The "easiest" solution would be to fix the upstream issue, i.e., do a PR to Clap's repository, but that I will only look at after this month,

Rolv-Apneseth commented:
Yeah couldn't find any crates for this situation.

I do think just copying the struct is better but to provide another option, what do you think of something like this:

```rust
pub struct Request<'source> {
    // HACK: keep compiler happy when enabling the cli feature flag
    #[cfg(feature = "cli")]
    #[clap(skip)]
    _lifetime: PhantomData<&'source ()>,
    /// (...docs)
    #[cfg(feature = "cli")]
    #[clap(short = 't', long, conflicts_with = "data", allow_hyphen_values(true))]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<Cow<'static, str>>,
    /// (...docs)
    #[cfg(not(feature = "cli"))]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<Cow<'source, str>>,
    /// (...docs)
    #[cfg(feature = "cli")]
    #[clap(short = 'd', long, conflicts_with = "text")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Data<'static>>,
    /// (...docs)
    #[cfg(not(feature = "cli"))]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Data<'source>>,
```

As for the upstream issue - I had a look but I believe it's beyond my skills. Needs quite a bit of familiarity with how their derive macros work, and the author seems to agree having assigned it that `E-hard` tag. If you want to hold this off until you have a look at making a PR yourself though, that would be fine by me.

jeertmans commented:
> e struct is better but to provide another option,

This is not a wrong idea, but the issue is that compiling with `cli` will then enforce `'static` lifetime everywhere. Which is something we want to avoid even in the CLI, when we read text from files with `ltrs check FILES...`

This is an edge case, but we don't want to have to keep a static lifetime reference to the content of each file, where we don't need it. So copying the struct is the only we to both have Clap happy and still be able to use non-static lifetime when desired.

> As for the upstream issue - I had a look but I believe it's beyond my skills. Needs quite a bit of familiarity with how their derive macros work, and the author seems to agree having assigned it that E-hard tag. If you want to hold this off until you have a look at making a PR yourself though, that would be fine by me.

I agree, this is not easy, especially as it involves proc macros!

Rolv-Apneseth commented:
So, it's not ideal, and I had to use the [lifetime crate](https://crates.io/crates/lifetime), but what do you think of the latest commit?

Perhaps it's better to leave those fields as `String` until `clap` allows us to avoid all the duplication?

jeertmans commented on src/api/check.rs:
```diff
@@ -963,7 +953,7 @@ impl Response {
 pub struct ResponseWithContext {
     /// Original text that was checked by LT.
-    pub text: Cow<'static, str>,
+    pub text: String,
```
Why does it have to be an owned string? If I am correct, the "context" is simply a source text, so should be `&'source str`, no?

Rolv-Apneseth replied:
I just had issues with lifetimes, in particular `check_multiple_and_join` in `server.rs` was a pain due to those tokio tasks that are spawned. However, I'm having a look at if it can maybe work with `IntoStatic`.

What is the match positions section in this code block supposed to do? It's causing a lifetime issue and I don't see the values being returned / used for anything:

```rust
impl<'source> From<ResponseWithContext<'source>> for Response {
    #[allow(clippy::needless_borrow)]
    fn from(mut resp: ResponseWithContext<'source>) -> Self {
        let iter: MatchPositions<'_, std::slice::IterMut<'_, Match>> = (&mut resp).into();

        for (line_number, line_offset, m) in iter {
            m.more_context = Some(MoreContext {
                line_number,
                line_offset,
            });
        }
        resp.response
    }
}
```

jeertmans replied:
Hum I see, probably using multiple tasks may conflict with lifetime.

The match `iter` is an iterator that will automatically count lines and line offsets, so we can then add this information to the `Match` object.

Rolv-Apneseth replied:
Thanks that helped, so that piece of code modifies the response with some additional context before returning it.

Have a look at the latest commit and see if you're happy with the changes as I needed some workarounds for more lifetime issues due to that block of code above.

I could improve it if we drop the implementation for `Iter` and just keep the one for `IterMut`, so that it doesn't need to be generic. Also, maybe the modification of `response` with that additional context could happen when `ResponseWithContext` is first created, rather than only when converting it to `Response`?

jeertmans replied:
Hi @Rolv-Apneseth! Your work looks very promising, but I have a paper deadline (next Thursday) that is taking most of my time, so I will (hopefully, if the deadline is not postponed) have time to review your work on next Friday :-)

Rolv-Apneseth replied:
Of course, no worries, you can take your time. And good luck!

Rolv-Apneseth replied:
Were you alright with these changes then? If so I can move on to the de-duplication of the `DataAnnotation` and `CliRequest` methods

Rolv-Apneseth replied:
Also - hope the paper went well

jeertmans commented on src/cli/check.rs:
```diff
@@ -112,10 +120,11 @@ impl ExecuteSubcommand for Command {
         // FILES
         for filename in self.filenames.iter() {
             let text = std::fs::read_to_string(filename)?;
+            let split_pattern = self.split_pattern.clone();
```
Is the clone necessary?

Rolv-Apneseth replied:
No it is not, sorry, must've been leftover from something I was trying. I'll remove that line.

jeertmans commented on src/cli/check.rs:
```diff
@@ -132,3 +141,373 @@ impl ExecuteSubcommand for Command {
+    /// Return a copy of the text within the request.
+    /// Call [`Request::try_get_text`] but panic on error.
+    ///
+    /// # Panics
+    ///
+    /// If both `self.text` and `self.data` are [`None`].
+    /// If any data annotation does not contain text or markup.
+    #[must_use]
+    pub fn get_text(&self) -> String {
+        self.try_get_text().unwrap()
+    }
+
+    /// Split this request into multiple, using [`split_len`] function to split
+    /// text.
+    ///
+    /// # Errors
+    ///
+    /// If `self.text` is none.
+    pub fn try_split(mut self, n: usize, pat: &str) -> Result<Vec<Request>> {
+        let text = mem::take(&mut self.text)
+            .ok_or_else(|| Error::InvalidRequest("missing text field".to_string()))?;
+
+        Ok(split_len(&text, n, pat)
+            .iter()
+            .map(|text_fragment| self.clone().with_text(text_fragment.to_string()).into())
+            .collect())
+    }
+
+    /// Split this request into multiple, using [`split_len`] function to split
+    /// text.
+    /// Call [`Request::try_split`] but panic on error.
+    ///
+    /// # Panics
+    ///
+    /// If `self.text` is none.
+    #[must_use]
+    pub fn split(self, n: usize, pat: &str) -> Vec<Request> {
+        self.try_split(n, pat).unwrap()
+    }
+}
```
I think all of this should be only implemented for `Request`, not `CliRequest`. You can simply have the CLI have the following logic:

```
cli_req = CliRequest::parse();
req: Request = cli_req.into();
req.some_function...
```

Rolv-Apneseth replied:
Alright, makes sense, I'll see if that works

jeertmans commented on src/cli/check.rs:
```diff
@@ -132,3 +141,373 @@ impl ExecuteSubcommand for Command {
+    pub fn new_markup<M: Into<String>>(markup: M) -> Self {
+        Self {
+            text: None,
+            markup: Some(markup.into()),
+            interpret_as: None,
+        }
+    }
+
+    /// Instantiate a new `CliDataAnnotation` with markup and its
+    /// interpretation.
+    #[inline]
+    #[must_use]
+    pub fn new_interpreted_markup<M: Into<String>, I: Into<String>>(
+        markup: M,
+        interpret_as: I,
+    ) -> Self {
+        Self {
+            interpret_as: Some(interpret_as.into()),
+            markup: Some(markup.into()),
+            text: None,
+        }
+    }
+
+    /// Return the text or markup within the data annotation.
+    ///
+    /// # Errors
+    ///
+    /// If this data annotation does not contain text or markup.
+    pub fn try_get_text(&self) -> Result<String> {
+        if let Some(ref text) = self.text {
+            Ok(text.clone())
+        } else if let Some(ref markup) = self.markup {
+            Ok(markup.clone())
+        } else {
+            Err(Error::InvalidDataAnnotation(format!(
+                "missing either text or markup field in {self:?}"
+            )))
+        }
+    }
+}
```
Same remark as for `CliRequest`

jeertmans reviewed (commented):
Hello @Rolv-Apneseth! It looks great, I just have a few comments that I left :-)

Please let me know what you think of it!

jeertmans commented on README.md:
```diff
@@ -154,13 +154,14 @@ languagetool-rust = "^2.1"
 
     let req = check::Request::default()
-        .with_text("Some phrase with a smal mistake".to_string());  // # codespell:ignore smal
+        .with_text(Cow::Borrowed("Some phrase with a smal mistake"));  // # codespell:ignore smal
```
`with_text` accepts anything that implements `Into<Cow<'_, str>>`, so passing a `&str` should be fine and more readable.
```suggestion
        .with_text("Some phrase with a smal mistake");  // # codespell:ignore smal
```

Rolv-Apneseth replied:
Yep makes sense

jeertmans commented on benches/benchmarks/check_texts.rs:
```diff
@@ -48,7 +50,7 @@ async fn check_text_split(text: &str) -> Response {
     let resps = join_all(lines.map(|line| {
         async {
-            let req = Request::default().with_text(line.to_string());
+            let req = Request::default().with_text(Cow::Owned(line.to_string()));
```
Same comment as above
```suggestion
            let req = Request::default().with_text(line.to_string());
```

jeertmans commented on benches/benchmarks/check_texts.rs:
```diff
@@ -29,12 +31,12 @@ async fn request_until_success(req: &Request, client: &ServerClient) -> Response
          it.",
     );
-    let req = Request::default().with_text(text.to_string());
+    let req = Request::default().with_text(Cow::Borrowed(text));
```
Same comment as above
```suggestion
    let req = Request::default().with_text(text);
```

jeertmans commented on benches/benchmarks/check_texts.rs:
```diff
@@ -29,12 +31,12 @@ async fn request_until_success(req: &Request, client: &ServerClient) -> Response
 
 #[tokio::main]
-async fn check_text_basic(text: &str) -> Response {
+async fn check_text_basic(text: &'static str) -> Response {
```
Is there any reason for this to be static?

Rolv-Apneseth replied:
This was only static because of `with_text` and `with_data` - I'll fix that

jeertmans commented on src/api/check.rs:
```diff
@@ -475,79 +491,51 @@ pub struct Request {
     /// Set the data to be checked and remove potential text field.
     #[must_use]
-    pub fn with_data(mut self, data: Data) -> Self {
+    pub fn with_data(mut self, data: Data<'static>) -> Self {
```
Why do we have static lifetime here? I thought this would be limited to the CLI API.

Rolv-Apneseth replied:
Not sure sorry, I'll remove it

jeertmans commented on src/api/check.rs:
```diff
@@ -1090,42 +1025,59 @@ impl ResponseWithContext {
-        self.text.push_str(other.text.as_str());
-        self.text_length += other.text_length;
+
+        self.text = Cow::Owned(self.text.into_owned() + &other.text);
```
Probably "nicer" to use https://doc.rust-lang.org/std/borrow/enum.Cow.html#method.to_mut instead, though this eventually does the same

Rolv-Apneseth replied:
I can change that yeah - never properly used `Cow` before to be honest, good to learn all these

jeertmans commented on src/api/server.rs:
```diff
@@ -595,7 +607,7 @@ mod tests {
     async fn test_server_check_text() {
         let client = ServerClient::from_env_or_default();
-        let req = Request::default().with_text("je suis une poupee".to_string());
+        let req = Request::default().with_text(Cow::Borrowed("je suis une poupee"));
```
```suggestion
        let req = Request::default().with_text("je suis une poupee");
```

jeertmans commented on tests/match_positions.rs:
```diff
@@ -7,7 +8,7 @@ macro_rules! test_match_positions {
 
             let client = ServerClient::from_env_or_default();
-            let req = check::Request::default().with_text($text.to_string());
+            let req = check::Request::default().with_text(Cow::Borrowed($text));
```
```suggestion
            let req = check::Request::default().with_text($text);
```

jeertmans reviewed (commented):
Hello @Rolv-Apneseth, I finally have some time to review this work properly :-)

I did a new pass, and it looks excellent! I suggested a few changes, especially for readability, and added two questions regarding two static lifetimes.

I see that most tests are passing, except a few ones, but didn't yet understand why those tests were failing.

jeertmans commented:
Hello @Rolv-Apneseth, sorry for lagging a bit behind those last weeks. I looked at failing test a bit in depth and I don't mind merging this PR.

We will take care of making those tests pass in the main PR :-)

Rolv-Apneseth commented:
Oh - I hadn't done the de-duplication yet - shall I do that and you can merge this branch again? I was waiting for a reply to this:

> Were you alright with these changes then? If so I can move on to the de-duplication of the DataAnnotation and CliRequest methods

jeertmans commented:
Oh, sorry! Well, given your large contribution to this repo, I consider adding you as a contributor so you can directly edit `v3`, etc.

Should I add you?

Rolv-Apneseth commented:
I mean, as long as you're comfortable with it. Personally I wouldn't trust strangers on the internet so fast haha

jeertmans commented:
> I mean, as long as you're comfortable with it. Personally I wouldn't trust strangers on the internet so fast haha

I know, but branch protection rules are good to prevent issues :-)

Rolv-Apneseth commented:
Fair enough yeah. Thanks for the invite, I'll push that deduplication to v3 (probably) later today

jeertmans commented:
Thanks! I plan to be more active on this repo myself, as I plan on using this combined with Typst to write my thesis, but this is unfortunately not a priority at the moment, and this is why I prefer to give you more freedom to work on this project while you are motivated, rather than to stale it forever haha
